File: demos/android/app/src/main/java/com/demo/rpcbridge/RpcBridgeServer.kt
/**
 * Kotlin server-side RPC runtime for the bridge protocol.
 *
 * Handles incoming frames from the WebView transport, manages stream state,
 * dispatches to service handlers, and sends response frames back. This is
 * the Kotlin equivalent of RpcServer in @rpc-bridge/core.
 *
 * Uses Kotlin coroutines and Flow for async stream processing.
 */
package com.demo.rpcbridge

import android.util.Log
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.launch

private const val TAG = "RpcBridgeServer"

// ---------------------------------------------------------------------------
// Handler type aliases
// ---------------------------------------------------------------------------

/** Unary: single request -> single response. */
typealias UnaryHandler = suspend (request: ByteArray) -> ByteArray

/** Server-streaming: single request -> stream of responses. */
typealias ServerStreamHandler = (request: ByteArray) -> Flow<ByteArray>

/** Client-streaming: stream of requests -> single response. */
typealias ClientStreamHandler = suspend (requests: Flow<ByteArray>) -> ByteArray

/** Bidi-streaming: stream of requests -> stream of responses. */
typealias BidiStreamHandler = (requests: Flow<ByteArray>) -> Flow<ByteArray>

// ---------------------------------------------------------------------------
// Method registration
// ---------------------------------------------------------------------------

/** Describes one registered RPC method with its handler. */
sealed class MethodRegistration(val name: String) {
    class Unary(name: String, val handler: UnaryHandler) : MethodRegistration(name)
    class ServerStream(name: String, val handler: ServerStreamHandler) : MethodRegistration(name)
    class ClientStream(name: String, val handler: ClientStreamHandler) : MethodRegistration(name)
    class BidiStream(name: String, val handler: BidiStreamHandler) : MethodRegistration(name)
}

/** Service registration grouping methods under a fully-qualified service name. */
data class ServiceRegistration(
    val name: String,
    val methods: Map<String, MethodRegistration>,
)

// ---------------------------------------------------------------------------
// Stream state tracking
// ---------------------------------------------------------------------------

private enum class StreamState {
    OPEN,
    HALF_CLOSED_REMOTE,
    CLOSED,
    CANCELLED,
}

/**
 * Tracks the server-side state of one logical RPC stream.
 *
 * Incoming messages are buffered in a Channel so handler coroutines can
 * consume them asynchronously via Flow.
 */
private class StreamInfo(
    val streamId: Int,
    val method: String,
) {
    var state: StreamState = StreamState.OPEN
    var sendSequence: Int = 0

    /** Channel for incoming MESSAGE payloads (client -> server). */
    val incoming: Channel<ByteArray> = Channel(Channel.UNLIMITED)

    /** Coroutine job running the handler, used for cancellation. */
    var handlerJob: Job? = null

    fun nextSendSequence(): Int = ++sendSequence
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Default flow control credits. Matches DEFAULT_INITIAL_CREDITS in the TS runtime. */
private const val DEFAULT_INITIAL_CREDITS = 16

/** Protocol version advertised by this implementation. */
private const val PROTOCOL_VERSION = 1

/** Human-readable implementation identifier for handshake. */
private const val IMPLEMENTATION_ID = "rpc-bridge-android-demo/0.1.0"

/** Default capabilities advertised by this implementation. */
private val DEFAULT_CAPABILITIES = listOf("flow_control", "deadline", "cancellation")

// ---------------------------------------------------------------------------
// Error codes (subset of gRPC status codes)
// ---------------------------------------------------------------------------

object RpcStatusCode {
    const val OK = 0
    const val CANCELLED = 1
    const val UNKNOWN = 2
    const val INVALID_ARGUMENT = 3
    const val UNIMPLEMENTED = 12
    const val INTERNAL = 13
}

// ---------------------------------------------------------------------------
// RpcBridgeServer
// ---------------------------------------------------------------------------

/**
 * Server runtime that processes RPC frames and dispatches to registered
 * service handlers. Typically bound to a [NativeBridgeTransport] which
 * bridges frames between the WebView and this server.
 *
 * @param scope     CoroutineScope for launching handler coroutines
 * @param sendFrame Callback to send an encoded RpcFrame back to the WebView
 */
class RpcBridgeServer(
    private val scope: CoroutineScope,
    private val sendFrame: (RpcFrame) -> Unit,
) {
    private val services = mutableMapOf<String, ServiceRegistration>()
    private val streams = mutableMapOf<Int, StreamInfo>()
    private var handshakeComplete = false

    // --- Public API ---

    /** Register a service with the server. */
    fun registerService(service: ServiceRegistration) {
        services[service.name] = service
        Log.d(TAG, "Registered service: ${service.name} (${service.methods.size} methods)")
    }

    /**
     * Handle an incoming frame from the transport layer.
     * This is the main entry point called by [NativeBridgeTransport].
     */
    fun handleFrame(frame: RpcFrame) {
        when (frame.type) {
            FrameType.HANDSHAKE -> handleHandshake(frame)
            FrameType.OPEN -> handleOpen(frame)
            FrameType.MESSAGE -> handleMessage(frame)
            FrameType.HALF_CLOSE -> handleHalfClose(frame)
            FrameType.CANCEL -> handleCancel(frame)
            FrameType.REQUEST_N -> handleRequestN(frame)
            else -> Log.d(TAG, "Ignoring unknown frame type ${frame.type} on stream ${frame.streamId}")
        }
    }

    /** Shut down the server and cancel all active streams. */
    fun close() {
        for ((_, stream) in streams) {
            stream.incoming.close()
            stream.handlerJob?.cancel()
        }
        streams.clear()
        Log.d(TAG, "Server closed")
    }

    // --- Handshake ---

    private fun handleHandshake(frame: RpcFrame) {
        Log.d(TAG, "Received handshake from peer: v${frame.protocolVersion}, " +
            "caps=[${frame.capabilities?.joinToString(",") ?: ""}], " +
            "impl=${frame.implementationId ?: "unknown"}")

        // Send our handshake response
        val response = createHandshakeFrame(
            protocolVersion = PROTOCOL_VERSION,
            capabilities = DEFAULT_CAPABILITIES,
            implementationId = IMPLEMENTATION_ID,
        )
        sendFrame(response)

        handshakeComplete = true
        Log.i(TAG, "Handshake complete")
    }

    // --- Stream: OPEN ---

    private fun handleOpen(frame: RpcFrame) {
        val method = frame.method
        if (method.isNullOrEmpty()) {
            sendError(frame.streamId, RpcStatusCode.INVALID_ARGUMENT, "Missing method name")
            return
        }

        // Parse "package.ServiceName/MethodName"
        val slashIdx = method.lastIndexOf('/')
        if (slashIdx < 0) {
            sendError(frame.streamId, RpcStatusCode.INVALID_ARGUMENT, "Invalid method format: $method")
            return
        }
        val serviceName = method.substring(0, slashIdx)
        val methodName = method.substring(slashIdx + 1)

        val service = services[serviceName]
        if (service == null) {
            sendError(frame.streamId, RpcStatusCode.UNIMPLEMENTED, "Unknown service: $serviceName")
            return
        }

        val registration = service.methods[methodName]
        if (registration == null) {
            sendError(frame.streamId, RpcStatusCode.UNIMPLEMENTED, "Unknown method: $method")
            return
        }

        // Create stream tracking
        val streamInfo = StreamInfo(frame.streamId, method)
        streams[frame.streamId] = streamInfo

        Log.d(TAG, "Stream ${frame.streamId} opened: $method")

        // Dispatch to the appropriate handler pattern
        dispatchMethod(streamInfo, registration)
    }

    // --- Stream: MESSAGE ---

    private fun handleMessage(frame: RpcFrame) {
        val stream = streams[frame.streamId]
        if (stream == null) {
            Log.w(TAG, "MESSAGE for unknown stream ${frame.streamId}")
            return
        }
        if (stream.state == StreamState.CANCELLED || stream.state == StreamState.CLOSED) {
            return
        }
        val payload = frame.payload ?: ByteArray(0)
        stream.incoming.trySend(payload)
    }

    // --- Stream: HALF_CLOSE ---

    private fun handleHalfClose(frame: RpcFrame) {
        val stream = streams[frame.streamId]
        if (stream == null) {
            Log.w(TAG, "HALF_CLOSE for unknown stream ${frame.streamId}")
            return
        }
        stream.state = StreamState.HALF_CLOSED_REMOTE
        stream.incoming.close() // Signal end-of-input to the handler
        Log.d(TAG, "Stream ${frame.streamId} half-closed by client")
    }

    // --- Stream: CANCEL ---

    private fun handleCancel(frame: RpcFrame) {
        val stream = streams[frame.streamId] ?: return
        Log.d(TAG, "Stream ${frame.streamId} cancelled by client")
        stream.state = StreamState.CANCELLED
        stream.incoming.close()
        stream.handlerJob?.cancel()
        streams.remove(frame.streamId)
    }

    // --- Stream: REQUEST_N (flow control) ---

    private fun handleRequestN(frame: RpcFrame) {
        // In this demo we don't block on send-side credits; we log and
        // continue. A production implementation would track credits and
        // suspend senders when exhausted.
        Log.d(TAG, "Stream ${frame.streamId}: granted ${frame.requestN} additional credits")
    }

    // --- Handler dispatch ---

    private fun dispatchMethod(stream: StreamInfo, registration: MethodRegistration) {
        stream.handlerJob = scope.launch {
            try {
                when (registration) {
                    is MethodRegistration.Unary ->
                        handleUnary(stream, registration.handler)
                    is MethodRegistration.ServerStream ->
                        handleServerStream(stream, registration.handler)
                    is MethodRegistration.ClientStream ->
                        handleClientStream(stream, registration.handler)
                    is MethodRegistration.BidiStream ->
                        handleBidiStream(stream, registration.handler)
                }
            } catch (e: Exception) {
                if (stream.state != StreamState.CANCELLED) {
                    Log.e(TAG, "Handler error for ${stream.method}: ${e.message}", e)
                    sendError(stream.streamId, RpcStatusCode.INTERNAL, e.message ?: "Internal error")
                }
            } finally {
                streams.remove(stream.streamId)
            }
        }
    }

    // --- Unary ---

    private suspend fun handleUnary(stream: StreamInfo, handler: UnaryHandler) {
        // Wait for the single request message
        val requestBytes = stream.incoming.receive()

        // Wait for HALF_CLOSE (client signals end of input)
        // For unary calls the client sends MESSAGE then HALF_CLOSE.
        // The channel will be closed by handleHalfClose.
        drainRemaining(stream)

        // Call handler
        val responseBytes = handler(requestBytes)

        // Send response
        sendFrame(
            createMessageFrame(stream.streamId, stream.nextSendSequence(), responseBytes)
        )

        // Close stream
        sendFrame(createCloseFrame(stream.streamId))
        stream.state = StreamState.CLOSED
    }

    // --- Server-streaming ---

    private suspend fun handleServerStream(stream: StreamInfo, handler: ServerStreamHandler) {
        // Wait for the single request message
        val requestBytes = stream.incoming.receive()
        drainRemaining(stream)

        // Call handler to get the response Flow
        val responses = handler(requestBytes)

        // Send each response message
        responses.collect { responseBytes ->
            if (stream.state == StreamState.CANCELLED) return@collect
            sendFrame(
                createMessageFrame(stream.streamId, stream.nextSendSequence(), responseBytes)
            )
        }

        // Close stream
        if (stream.state != StreamState.CANCELLED) {
            sendFrame(createCloseFrame(stream.streamId))
            stream.state = StreamState.CLOSED
        }
    }

    // --- Client-streaming ---

    private suspend fun handleClientStream(stream: StreamInfo, handler: ClientStreamHandler) {
        // Grant initial credits so the client can start sending
        sendFrame(createRequestNFrame(stream.streamId, DEFAULT_INITIAL_CREDITS))

        // Build a Flow from the incoming channel
        val requestFlow = channelToFlow(stream)

        // Call handler
        val responseBytes = handler(requestFlow)

        // Send response
        sendFrame(
            createMessageFrame(stream.streamId, stream.nextSendSequence(), responseBytes)
        )

        // Close stream
        sendFrame(createCloseFrame(stream.streamId))
        stream.state = StreamState.CLOSED
    }

    // --- Bidi-streaming ---

    private suspend fun handleBidiStream(stream: StreamInfo, handler: BidiStreamHandler) {
        // Grant initial credits so the client can start sending
        sendFrame(createRequestNFrame(stream.streamId, DEFAULT_INITIAL_CREDITS))

        // Build a Flow from the incoming channel
        val requestFlow = channelToFlow(stream)

        // Call handler to get the response Flow
        val responses = handler(requestFlow)

        // Send each response message
        responses.collect { responseBytes ->
            if (stream.state == StreamState.CANCELLED) return@collect
            sendFrame(
                createMessageFrame(stream.streamId, stream.nextSendSequence(), responseBytes)
            )
        }

        // Send HALF_CLOSE from server side, then CLOSE
        if (stream.state != StreamState.CANCELLED) {
            sendFrame(createHalfCloseFrame(stream.streamId))
            sendFrame(createCloseFrame(stream.streamId))
            stream.state = StreamState.CLOSED
        }
    }

    // --- Helpers ---

    /**
     * Convert a stream's incoming Channel into a Flow for handler consumption.
     * Automatically replenishes flow-control credits as messages are consumed.
     */
    private fun channelToFlow(stream: StreamInfo): Flow<ByteArray> = flow {
        var consumed = 0
        for (bytes in stream.incoming) {
            emit(bytes)
            consumed++
            // Replenish credits when we've consumed a quarter of the window
            if (consumed >= DEFAULT_INITIAL_CREDITS / 4) {
                sendFrame(createRequestNFrame(stream.streamId, consumed))
                consumed = 0
            }
        }
    }

    /**
     * Drain any remaining messages from the incoming channel.
     * Used after receiving the expected request in unary/server-streaming patterns
     * to consume the HALF_CLOSE signal.
     */
    private suspend fun drainRemaining(stream: StreamInfo) {
        // The channel will be closed when HALF_CLOSE arrives.
        // We consume and discard any extra messages (there shouldn't be any
        // for unary/server-streaming, but this keeps the protocol robust).
        for (extra in stream.incoming) {
            Log.w(TAG, "Stream ${stream.streamId}: unexpected extra message after request")
        }
    }

    /** Send an ERROR frame and clean up the stream. */
    private fun sendError(streamId: Int, code: Int, message: String) {
        try {
            sendFrame(createErrorFrame(streamId, code, message))
        } catch (e: Exception) {
            Log.e(TAG, "Failed to send error frame: ${e.message}")
        }
        streams.remove(streamId)
    }
}

File: demos/ios/RPCBridgeDemo/Sources/RpcBridgeServer.swift
// RpcBridgeServer.swift
// RPCBridgeDemo
//
// Server-side RPC runtime for iOS. Handles incoming frames from the
// WKWebView client, manages stream state, dispatches to service handlers,
// and sends response frames back through the transport.
//
// This mirrors the TypeScript RpcServer in packages/rpc-core/src/server.ts,
// implementing the same frame-level protocol:
// - HANDSHAKE: Protocol version negotiation
// - OPEN: Stream creation and method dispatch
// - MESSAGE: Payload delivery with flow control
// - HALF_CLOSE: End of client messages
// - CLOSE: Successful stream completion
// - ERROR: Stream-level error signaling
// - CANCEL: Stream cancellation
// - REQUEST_N: Credit-based flow control

import Foundation

// MARK: - Protocol Version Constants

/// Current protocol version. Must match the TypeScript side.
private let currentProtocolVersion: UInt32 = 1

/// Implementation identifier for this Swift runtime.
private let swiftImplementationId = "rpc-bridge-ios-demo/1.0.0"

/// Default capabilities advertised by this implementation.
/// Must match the capabilities in the TypeScript handshake module.
private let defaultCapabilities = ["flow_control", "deadline", "cancellation"]

/// Default initial flow control credits granted to each stream.
private let defaultInitialCredits: UInt32 = 16

// MARK: - Stream State

/// Possible states for a server-side stream.
/// Follows the same state machine as the TypeScript Stream class.
enum StreamState: Sendable {
    case idle
    case open
    case halfClosedRemote   // Client sent HALF_CLOSE (no more client messages)
    case halfClosedLocal    // Server sent HALF_CLOSE (no more server messages)
    case halfClosedBoth
    case closed
    case error
    case cancelled
}

// MARK: - Server Stream

/// Manages the state and message buffering for a single server-side stream.
/// Each RPC call creates one ServerStream instance that tracks its lifecycle
/// from OPEN through CLOSE/ERROR/CANCEL.
final class ServerStream: @unchecked Sendable {
    let streamId: UInt32
    private(set) var state: StreamState = .idle
    private var sendSequence: UInt32 = 0
    private var sendCredits: UInt32 = 0

    /// Continuation for feeding incoming messages to the async stream consumer.
    private var messageContinuation: AsyncStream<Data>.Continuation?

    /// The async stream that service handlers consume to receive client messages.
    let messages: AsyncStream<Data>

    /// Task managing the stream handler, used for cancellation.
    var handlerTask: Task<Void, Never>?

    init(streamId: UInt32) {
        self.streamId = streamId
        var continuation: AsyncStream<Data>.Continuation!
        self.messages = AsyncStream<Data> { c in
            continuation = c
        }
        self.messageContinuation = continuation
        // Grant initial send credits so the server can send responses
        self.sendCredits = defaultInitialCredits
    }

    /// Transition to open state.
    func open() {
        state = .open
    }

    /// Get and increment the send sequence number.
    func nextSendSequence() -> UInt32 {
        sendSequence += 1
        return sendSequence
    }

    /// Push an incoming message payload from the client.
    func pushMessage(_ data: Data) {
        messageContinuation?.yield(data)
    }

    /// Signal that the client will send no more messages.
    func pushEnd() {
        messageContinuation?.finish()
        messageContinuation = nil

        if state == .halfClosedLocal {
            state = .halfClosedBoth
        } else {
            state = .halfClosedRemote
        }
    }

    /// Signal a client-side error on this stream.
    func pushError() {
        messageContinuation?.finish()
        messageContinuation = nil
        state = .error
    }

    /// Cancel this stream.
    func cancel() {
        messageContinuation?.finish()
        messageContinuation = nil
        state = .cancelled
        handlerTask?.cancel()
    }

    /// Mark this stream as closed.
    func close() {
        messageContinuation?.finish()
        messageContinuation = nil
        state = .closed
    }

    /// Add send credits (called when REQUEST_N is received from the client).
    func addSendCredits(_ n: UInt32) {
        sendCredits += n
    }

    /// Consume one send credit. Returns true if a credit was available.
    func consumeSendCredit() -> Bool {
        if sendCredits > 0 {
            sendCredits -= 1
            return true
        }
        return false
    }
}

// MARK: - Frame Send Callback

/// Callback type for sending frames back to the WKWebView client.
/// The transport layer provides this closure when setting up the server.
typealias FrameSender = @Sendable (RpcFrame) -> Void

// MARK: - RpcBridgeServer

/// The server runtime that receives frames from the web client, dispatches
/// RPC calls to registered service handlers, and sends response frames back.
///
/// Usage:
/// 1. Create the server with a frame sender callback
/// 2. Register the HelloBridgeService implementation
/// 3. Feed incoming frames via `handleFrame(_:)`
///
/// The server manages the full RPC lifecycle including handshake negotiation,
/// stream creation, method dispatch, flow control, and error handling.
final class RpcBridgeServer: @unchecked Sendable {

    /// Callback for sending frames to the client.
    private let sendFrame: FrameSender

    /// Active streams keyed by stream ID.
    private var streams: [UInt32: ServerStream] = [:]

    /// Lock for thread-safe stream map access.
    private let lock = NSLock()

    /// Whether the handshake has been completed.
    private var handshakeCompleted = false

    /// The registered service implementation.
    private let service: HelloBridgeServiceProvider

    /// Fully qualified service name matching the proto package + service name.
    /// This must match what the TypeScript client sends in OPEN frames.
    private let serviceName = "demo.hello.v1.HelloBridgeService"

    /// Optional logging function for debugging.
    var log: ((String) -> Void)?

    // MARK: - Initialization

    init(service: HelloBridgeServiceProvider, sendFrame: @escaping FrameSender) {
        self.service = service
        self.sendFrame = sendFrame
    }

    // MARK: - Frame Handling

    /// Process an incoming frame from the web client.
    /// This is the main entry point called by the transport layer.
    /// Routes frames to the appropriate handler based on frame type.
    func handleFrame(_ frame: RpcFrame) {
        switch frame.type {
        case .handshake:
            handleHandshake(frame)

        case .open:
            handleOpen(frame)

        case .message:
            handleMessage(frame)

        case .halfClose:
            handleHalfClose(frame)

        case .cancel:
            handleCancel(frame)

        case .requestN:
            handleRequestN(frame)

        case .error:
            handleClientError(frame)

        case .close, .unspecified:
            // CLOSE from client is unusual but harmless; UNSPECIFIED is ignored
            log?("[Server] Ignoring frame type: \(frame.type) on stream \(frame.streamId)")
        }
    }

    // MARK: - Handshake

    /// Handle the client's HANDSHAKE frame and send our response.
    /// Negotiates protocol version (minimum of both sides) and
    /// intersects capabilities for feature negotiation.
    private func handleHandshake(_ frame: RpcFrame) {
        let peerVersion = frame.protocolVersion ?? 1
        let negotiatedVersion = min(currentProtocolVersion, peerVersion)

        log?("[Server] Handshake received: v\(peerVersion), peer=\(frame.implementationId ?? "unknown")")

        // Send our handshake response
        let response = createHandshakeFrame(
            protocolVersion: currentProtocolVersion,
            capabilities: defaultCapabilities,
            implementationId: swiftImplementationId
        )
        sendFrame(response)
        handshakeCompleted = true

        log?("[Server] Handshake complete: negotiated v\(negotiatedVersion)")
    }

    // MARK: - Stream Open

    /// Handle an OPEN frame to start a new RPC stream.
    /// Parses the method name, validates the service, and dispatches
    /// to the appropriate handler based on the method type.
    private func handleOpen(_ frame: RpcFrame) {
        guard let method = frame.method, !method.isEmpty else {
            sendError(streamId: frame.streamId, code: RpcStatusCode.invalidArgument, message: "Missing method name")
            return
        }

        // Parse method: "package.ServiceName/MethodName"
        guard let slashIdx = method.lastIndex(of: "/") else {
            sendError(streamId: frame.streamId, code: RpcStatusCode.invalidArgument, message: "Invalid method format: \(method)")
            return
        }

        let svcName = String(method[method.startIndex..<slashIdx])
        let methodName = String(method[method.index(after: slashIdx)...])

        // Verify the service name matches
        guard svcName == serviceName else {
            sendError(streamId: frame.streamId, code: RpcStatusCode.unimplemented, message: "Unknown service: \(svcName)")
            return
        }

        // Create the server-side stream
        let stream = ServerStream(streamId: frame.streamId)
        stream.open()
        lock.lock()
        streams[frame.streamId] = stream
        lock.unlock()

        log?("[Server] Stream \(frame.streamId) opened for method: \(method)")

        // Dispatch to the appropriate handler based on method name
        let task = Task { [weak self] in
            guard let self else { return }
            await self.dispatchMethod(
                stream: stream,
                methodName: methodName,
                methodType: frame.methodType ?? .unary
            )
        }
        stream.handlerTask = task
    }

    /// Dispatch an RPC call to the correct service handler method.
    private func dispatchMethod(
        stream: ServerStream,
        methodName: String,
        methodType: MethodType
    ) async {
        do {
            switch methodName {
            case "SayHello":
                try await handleSayHello(stream: stream)

            case "WatchGreeting":
                try await handleWatchGreeting(stream: stream)

            case "CollectNames":
                try await handleCollectNames(stream: stream)

            case "Chat":
                try await handleChat(stream: stream)

            default:
                sendError(
                    streamId: stream.streamId,
                    code: RpcStatusCode.unimplemented,
                    message: "Unknown method: \(serviceName)/\(methodName)"
                )
                removeStream(stream.streamId)
                return
            }
        } catch is CancellationError {
            // Stream was cancelled; nothing to do
            log?("[Server] Stream \(stream.streamId) handler cancelled")
        } catch {
            // Send error frame for unhandled errors
            let message = String(describing: error)
            sendError(streamId: stream.streamId, code: RpcStatusCode.internal, message: message)
        }

        removeStream(stream.streamId)
    }

    // MARK: - Unary Handler: SayHello

    /// Handle the SayHello unary RPC.
    /// Waits for the client's single request message, calls the service,
    /// then sends the single response followed by CLOSE.
    private func handleSayHello(stream: ServerStream) async throws {
        // Wait for the single request message
        let requestData = try await collectUnaryRequest(stream: stream)
        let request = HelloRequest.decode(from: requestData)

        // Call the service handler
        let response = try await service.sayHello(request)

        // Send response MESSAGE
        let seq = stream.nextSendSequence()
        let msgFrame = createMessageFrame(
            streamId: stream.streamId,
            sequence: seq,
            payload: response.encode()
        )
        sendFrame(msgFrame)

        // Send CLOSE to complete the stream
        sendFrame(createCloseFrame(streamId: stream.streamId))
        stream.close()

        log?("[Server] SayHello completed on stream \(stream.streamId)")
    }

    // MARK: - Server Streaming Handler: WatchGreeting

    /// Handle the WatchGreeting server-streaming RPC.
    /// Waits for the client's single request, then streams multiple
    /// response events with flow control.
    private func handleWatchGreeting(stream: ServerStream) async throws {
        // Wait for the single request message
        let requestData = try await collectUnaryRequest(stream: stream)
        let request = GreetingStreamRequest.decode(from: requestData)

        // Call the service handler to get the response stream
        let responses = service.watchGreeting(request)

        // Send each response with flow control
        for try await event in responses {
            if stream.state == .cancelled { return }
            try Task.checkCancellation()

            let seq = stream.nextSendSequence()
            let msgFrame = createMessageFrame(
                streamId: stream.streamId,
                sequence: seq,
                payload: event.encode()
            )
            sendFrame(msgFrame)
        }

        // Send CLOSE to complete the stream
        sendFrame(createCloseFrame(streamId: stream.streamId))
        stream.close()

        log?("[Server] WatchGreeting completed on stream \(stream.streamId)")
    }

    // MARK: - Client Streaming Handler: CollectNames

    /// Handle the CollectNames client-streaming RPC.
    /// Sends initial REQUEST_N credits, feeds client messages to the
    /// service handler via AsyncStream, then sends the single response.
    private func handleCollectNames(stream: ServerStream) async throws {
        // Grant initial credits so the client can start sending
        sendFrame(createRequestNFrame(streamId: stream.streamId, n: defaultInitialCredits))

        // Create a typed async stream from the raw message stream
        let typedRequests = AsyncStream<CollectNamesRequest> { continuation in
            Task {
                var received: UInt32 = 0
                for await data in stream.messages {
                    let request = CollectNamesRequest.decode(from: data)
                    continuation.yield(request)

                    // Replenish credits when we've consumed a portion
                    received += 1
                    if received % (defaultInitialCredits / 4) == 0 {
                        self.sendFrame(createRequestNFrame(streamId: stream.streamId, n: defaultInitialCredits))
                    }
                }
                continuation.finish()
            }
        }

        // Call the service handler
        let response = try await service.collectNames(typedRequests)

        // Send response MESSAGE
        let seq = stream.nextSendSequence()
        let msgFrame = createMessageFrame(
            streamId: stream.streamId,
            sequence: seq,
            payload: response.encode()
        )
        sendFrame(msgFrame)

        // Send CLOSE to complete the stream
        sendFrame(createCloseFrame(streamId: stream.streamId))
        stream.close()

        log?("[Server] CollectNames completed on stream \(stream.streamId)")
    }

    // MARK: - Bidi Streaming Handler: Chat

    /// Handle the Chat bidirectional streaming RPC.
    /// Sends initial REQUEST_N credits, then processes client messages
    /// and server responses concurrently.
    private func handleChat(stream: ServerStream) async throws {
        // Grant initial credits so the client can start sending
        sendFrame(createRequestNFrame(streamId: stream.streamId, n: defaultInitialCredits))

        // Create a typed async stream from the raw message stream
        let typedRequests = AsyncStream<ChatMessage> { continuation in
            Task {
                var received: UInt32 = 0
                for await data in stream.messages {
                    let msg = ChatMessage.decode(from: data)
                    continuation.yield(msg)

                    // Replenish credits periodically
                    received += 1
                    if received % (defaultInitialCredits / 4) == 0 {
                        self.sendFrame(createRequestNFrame(streamId: stream.streamId, n: defaultInitialCredits))
                    }
                }
                continuation.finish()
            }
        }

        // Call the service handler to get the response stream
        let responses = service.chat(typedRequests)

        // Send each response event
        for try await event in responses {
            if stream.state == .cancelled { return }
            try Task.checkCancellation()

            let seq = stream.nextSendSequence()
            let msgFrame = createMessageFrame(
                streamId: stream.streamId,
                sequence: seq,
                payload: event.encode()
            )
            sendFrame(msgFrame)
        }

        // Send HALF_CLOSE from server side, then CLOSE
        sendFrame(createHalfCloseFrame(streamId: stream.streamId))
        sendFrame(createCloseFrame(streamId: stream.streamId))
        stream.close()

        log?("[Server] Chat completed on stream \(stream.streamId)")
    }

    // MARK: - Message/HalfClose/Cancel/RequestN/Error Handlers

    /// Handle an incoming MESSAGE frame by routing it to the appropriate stream.
    private func handleMessage(_ frame: RpcFrame) {
        lock.lock()
        let stream = streams[frame.streamId]
        lock.unlock()

        guard let stream else {
            log?("[Server] Received MESSAGE for unknown stream \(frame.streamId)")
            return
        }

        stream.pushMessage(frame.payload ?? Data())
    }

    /// Handle a HALF_CLOSE frame indicating the client is done sending.
    private func handleHalfClose(_ frame: RpcFrame) {
        lock.lock()
        let stream = streams[frame.streamId]
        lock.unlock()

        guard let stream else {
            log?("[Server] Received HALF_CLOSE for unknown stream \(frame.streamId)")
            return
        }

        stream.pushEnd()
        log?("[Server] Stream \(frame.streamId) half-closed by client")
    }

    /// Handle a CANCEL frame requesting stream cancellation.
    private func handleCancel(_ frame: RpcFrame) {
        lock.lock()
        let stream = streams[frame.streamId]
        lock.unlock()

        guard let stream else {
            log?("[Server] Received CANCEL for unknown stream \(frame.streamId)")
            return
        }

        stream.cancel()
        removeStream(frame.streamId)
        log?("[Server] Stream \(frame.streamId) cancelled by client")
    }

    /// Handle a REQUEST_N frame granting additional send credits.
    private func handleRequestN(_ frame: RpcFrame) {
        lock.lock()
        let stream = streams[frame.streamId]
        lock.unlock()

        guard let stream else { return }
        stream.addSendCredits(frame.requestN ?? 0)
    }

    /// Handle an ERROR frame from the client.
    private func handleClientError(_ frame: RpcFrame) {
        lock.lock()
        let stream = streams[frame.streamId]
        lock.unlock()

        guard let stream else { return }

        log?("[Server] Client error on stream \(frame.streamId): \(frame.errorMessage ?? "unknown")")
        stream.pushError()
        removeStream(frame.streamId)
    }

    // MARK: - Helpers

    /// Collect the single request message for a unary or server-streaming RPC.
    /// Waits for the first MESSAGE frame, which contains the request payload.
    private func collectUnaryRequest(stream: ServerStream) async throws -> Data {
        for await data in stream.messages {
            return data
        }
        throw RpcBridgeError.missingRequest
    }

    /// Send an error frame to the client.
    private func sendError(streamId: UInt32, code: UInt32, message: String) {
        let frame = createErrorFrame(
            streamId: streamId,
            errorCode: code,
            errorMessage: message
        )
        sendFrame(frame)
    }

    /// Remove a stream from the active streams map.
    private func removeStream(_ streamId: UInt32) {
        lock.lock()
        streams.removeValue(forKey: streamId)
        lock.unlock()
    }

    /// Cancel all active streams. Called when the transport is torn down.
    func cancelAll() {
        lock.lock()
        let allStreams = Array(streams.values)
        streams.removeAll()
        lock.unlock()

        for stream in allStreams {
            stream.cancel()
        }
    }
}

// MARK: - Errors

/// Errors specific to the RPC bridge server.
enum RpcBridgeError: Error, CustomStringConvertible {
    case missingRequest
    case unknownMethod(String)
    case transportClosed

    var description: String {
        switch self {
        case .missingRequest:
            return "Expected request message but stream ended"
        case .unknownMethod(let method):
            return "Unknown method: \(method)"
        case .transportClosed:
            return "Transport is closed"
        }
    }
}
